Faster ChorusNotes handling for LfMerge

rmunn opened the pull request:
Goes along with https://github.com/sillsdev/LfMerge/pull/338, which will be mergeable once this PR is merged and a new NuGet package is created with the LfMergeBridge code changes.

This PR moves several POCO classes from LfMerge into LfMergeBridge. This allows these classes to be passed by reference into the ChorusNotes action handler, saving two large JSON serialization steps, as suggested in https://github.com/sillsdev/LfMerge/issues/329.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/flexbridge/401)
<!-- Reviewable:end -->

rmunn committed:
Move some LfMerge model classes into LfMergeBridge

This will allow these classes to be passed by reference into the
ChorusNotes action handler, saving two large JSON serialization steps.

rmunn committed:
Allow passing comments to/from LfMerge w/o JSON

We use the ConditionalWeakTable class, designed to allow compilers to
attach metadata to objects, as an extra input/output mechanism that
doesn't require serializing the comments and replies to JSON strings.
This will save quite a lot of RAM when doing a Send/Receive of a project
with lots and lots of comments.

This implements the LfMergeBridge side of the process; a corresponding
change will be needed in LfMerge.

rmunn committed:
Allow FLExBridge to compile on Linux

Linux requires GenerateResourceUsePreserializedResources to be set to
true before `dotnet build` will allow non-string resources to be
compiled into assemblies.

rmunn commented:
About to rebase on top of `develop` now that the CI signing PR has been merged.

rmunn committed:
Split ExtraData into input and output

ConditionalWeakTable only allows one object per key, so it's better to
use separate tables for input and output rather than reusing a single
ExtraData table for both input and output.

rmunn committed:
Slightly improve LfComment creation syntax

megahirt commented on src/LfMergeBridge/LanguageForgeGetChorusNotesActionHandler.cs:
```diff
@@ -37,90 +38,125 @@ internal sealed class LanguageForgeGetChorusNotesActionHandler : IBridgeActionTy
+			else
+			{
+				// Older version of LfMerge, serialize (likely slow)
+				var serializedComments = new StringBuilder("New comments not yet in LF: ");
+				serializedComments.Append(JsonConvert.SerializeObject(lfComments));
+				LfMergeBridgeUtilities.AppendLineToSomethingForClient(ref somethingForClient, serializedComments.ToString());
+
+				var serializedReplies = new StringBuilder("New replies on comments already in LF: ");
+				serializedReplies.Append(JsonConvert.SerializeObject(lfReplies));
+				LfMergeBridgeUtilities.AppendLineToSomethingForClient(ref somethingForClient, serializedReplies.ToString());
 
-			var serializedStatusChanges = new StringBuilder("New status changes on comments already in LF: ");
-			serializedStatusChanges.Append(JsonConvert.SerializeObject(lfStatusChanges));
-			LfMergeBridgeUtilities.AppendLineToSomethingForClient(ref somethingForClient, serializedStatusChanges.ToString());
+				var serializedStatusChanges = new StringBuilder("New status changes on comments already in LF: ");
+				serializedStatusChanges.Append(JsonConvert.SerializeObject(lfStatusChanges));
+				LfMergeBridgeUtilities.AppendLineToSomethingForClient(ref somethingForClient, serializedStatusChanges.ToString());
+			}
```
I clarified this in person, and I don't think we want to keep the old implementation around.  We are switching to the new implementation that avoids string serialization of comments.  This is an opportunity to reduce complexity in our code.

rmunn replied:
Done in commit b572aa50.

rmunn committed:
No longer accept old serializedComments API

This simplifies the PR immensely, as we no longer need to deal with two
different versions (old and new) of the API.

megahirt reviewed (approved):
The worked to demonstrate that this new code works is complicated, so I won't verify this personally.  Robin has conveyed to me in person how this fits together and so I approve of this PR.  Looking forward to seeing it integrated in time!

rmunn commented:
I'm a bit nervous about having added GenerateResourceUsePreserializedResources to so many .csproj files; a cursory reading of https://www.github.com/dotnet/msbuild/issues/4704 suggests that it might cause the build to fail for net461 unless the System.Resources.Extensions package is also referenced by the .csproj file. It builds correctly without that package reference on my machine, though, and those reports are several years old, so hopefully that is not a problem any more. But if there are errors building for net461, I will open another PR to either add the System.Resources.Extensions NuGet package or else remove the GenerateResourceUsePreserializedResources property.

Noting this now because I'm about to merge this PR, and I want to make sure I can easily find the solution if it turns out merging the PR breaks the build of the `develop` branch.
